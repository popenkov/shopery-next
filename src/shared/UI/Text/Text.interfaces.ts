export enum TextTheme {
  PRIMARY = 'primary',
  INVERTED = 'inverted',
  ERROR = 'error',
}

export enum TextAlign {
  RIGHT = 'right',
  LEFT = 'left',
  CENTER = 'center',
}

export enum TextWeight {
  LIGHT = 'light',
  REGULAR = 'regular',
  MEDIUM = 'medium',
  SEMIBOLD = 'semibold',
  BOLD = 'bold',
}

export enum TextVariant {
  SEARCH_FIELD = 'search-field',
  DISPLAY_1 = 'display_1',
  HEADING_2 = 'heading_2',
  HEADING_3 = 'heading_3',
  HEADING_4 = 'heading_4',
  HEADING_5 = 'heading_5',
  BODY_XXL = 'body_xxl',
  BODY_XL = 'body_xl',
  BODY_L = 'body_l',
  BODY_M = 'body_m',
  BODY_S = 'body_s',
  BODY_XS = 'body_xs',
  BODY_TINY = 'body_tiny',
}
