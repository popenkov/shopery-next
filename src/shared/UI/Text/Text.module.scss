@import '@/shared/styles/mixins/text-styles.scss';

.text {
  &.error {
    color: var(--front-primary);
  }

  &.inverted {
    color: var(--front-primary-inversed);
  }

  &.left {
    text-align: left;
  }

  &.right {
    text-align: right;
  }

  &.center {
    text-align: center;
  }

  &.search_field {
    @include search-field;
  }

  &.display_1 {
    @include display-1;
  }

  &.heading_1 {
    @include heading-1;
  }

  &.heading_2 {
    @include heading-2;
  }

  &.heading_3 {
    @include heading-3;
  }

  &.heading_4 {
    @include heading-4;
  }

  &.heading_5 {
    @include heading-5;
  }

  &.body_xxl {
    @include body-xxl;
  }

  &.body_xl {
    @include body-xl;
  }

  &.body_l {
    @include body-l;
  }

  &.body_m {
    @include body-m;
  }

  &.body_s {
    @include body-s;
  }

  &.body_xs {
    @include body-xs;
  }

  &.body_tiny {
    @include body-tiny;
  }

  &.light {
    font-weight: 300;
  }

  &.medium {
    font-weight: 500;
  }

  &.semibold {
    font-weight: 600;
  }

  &.bold {
    font-weight: 700;
  }
}
