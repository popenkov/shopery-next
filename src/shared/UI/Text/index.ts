export { Text } from './Text';
export {
  TextTheme,
  TextAlign,
  TextWeight,
  TextVariant,
} from './Text.interfaces';
