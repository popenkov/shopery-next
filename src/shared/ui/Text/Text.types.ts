export type TTextTheme = 'primary' | 'inverted' | 'error';

export type TTextAlign = 'right' | 'left' | 'center';

export type TTextWeight = 'light' | 'regular' | 'medium' | 'semibold' | 'bold';

export type TTextVariant =
  | 'search-field'
  | 'display_1'
  | 'heading_1'
  | 'heading_2'
  | 'heading_3'
  | 'heading_4'
  | 'heading_5'
  | 'body_xxl'
  | 'body_xl'
  | 'body_l'
  | 'body_m'
  | 'body_s'
  | 'body_xs'
  | 'body_tiny';
