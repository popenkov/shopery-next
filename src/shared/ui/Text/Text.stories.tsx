import type { Meta, StoryObj } from '@storybook/react';

import { Text } from './Text';

const meta = {
    title: 'shared/Text',
    component: Text,
    parameters: {
        layout: 'centered',
    },
    tags: ['autodocs'],
    argTypes: {
        theme: ['primary', 'inverted', 'error'],
        align: ['right', 'left', 'center'],
        variant: [
            'search-field',
            'display_1',
            'heading_1',
            'heading_2',
            'heading_3',
            'heading_4',
            'heading_5',
            'body_xxl',
            'body_xl',
            'body_l',
            'body_m',
            'body_s',
            'body_xs',
            'body_tiny',
        ],
        weight: ['light', 'regular', 'medium', 'semibold', 'bold'],
        children: {
            controls: 'Text',
        },
    },
} satisfies Meta<typeof Text>;

export default meta;
type Story = StoryObj<typeof Text>;

export const TextStory: Story = {
    args: {
        children: 'Text',
        theme: 'primary',
        size: 'small',
    },
};
