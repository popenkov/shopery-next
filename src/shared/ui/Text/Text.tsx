import { memo } from 'react';
import type { ComponentProps, ElementType, ReactNode } from 'react';
import cn from 'classnames';

import cls from './Text.module.scss';
import { TextAlign, TextTheme, TextVariant, TextWeight } from './Text.types';

interface TextOwnProps<E extends ElementType = ElementType> {
  children: ReactNode;
  className?: string;
  href?: string;
  title?: string;
  text?: string;
  theme?: TextTheme;
  align?: TextAlign;
  variant?: TextVariant;
  weight?: TextWeight;
  as?: E;
}

const DEFAULT_ELEMENT: ElementType = 'p';

export type TextProps<E extends ElementType> = TextOwnProps<E> &
  Omit<ComponentProps<E>, keyof TextOwnProps>;

const T = <E extends ElementType = typeof DEFAULT_ELEMENT>(props: TextProps<E>) => {
  const {
    className,
    children,
    as,
    theme = 'primary',
    align = 'left',
    variant = 'body_m',
    weight = 'regular',
  } = props;

  const Element = as || DEFAULT_ELEMENT;

  const mods = {
    [cls[theme]]: true,
    [cls[variant]]: true,
    [cls[align]]: true,
    [cls[weight]]: true,
  };

  return <Element className={cn(cls.text, className, mods)}>{children}</Element>;
};

export const Text = memo(T);

Text.displayName = 'Text';
