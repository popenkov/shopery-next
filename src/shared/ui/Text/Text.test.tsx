import React from 'react';
import { render } from '@testing-library/react';

import { Text } from './Text';

describe('Text component', () => {
  it('renders correctly', () => {
    const { container } = render(<Text>Text content</Text>);
    expect(container).toMatchSnapshot();
  });

  it('renders correct class names', () => {
    const { container } = render(
      <Text theme="primary" variant="body_m" align="left" weight="regular">
        Text content
      </Text>,
    );
    expect(container.querySelector('.text')).toBeInTheDocument();
    expect(container.querySelector('.primary')).toBeInTheDocument();
    expect(container.querySelector('.body_m')).toBeInTheDocument();
    expect(container.querySelector('.left')).toBeInTheDocument();
    expect(container.querySelector('.regular')).toBeInTheDocument();
  });

  it('renders correct text', () => {
    const { getByText } = render(<Text>Text content</Text>);
    expect(getByText('Text content')).toBeInTheDocument();
  });

  it('renders correct element type', () => {
    const { container } = render(<Text as="span">Text content</Text>);
    expect(container.querySelector('span')).toBeInTheDocument();
  });

  it('renders correct href attribute', () => {
    const { container } = render(
      <Text href="https://example.com" as="a">
        Text content
      </Text>,
    );
    expect(container.querySelector('a')).toHaveAttribute('href', 'https://example.com');
  });

  it('renders correct title attribute', () => {
    const { container } = render(<Text title="Title">Text content</Text>);
    expect(container.querySelector('p')).toHaveAttribute('title', 'Title');
  });
});
