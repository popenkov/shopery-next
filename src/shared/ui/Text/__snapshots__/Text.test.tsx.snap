// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text component renders correctly 1`] = `
<div>
  <p
    class="text primary body_m left regular"
  >
    Text content
  </p>
</div>
`;
